File: prototype/TAT-C/matlab/test/SystemTest_Analysis.m
%% This is a usage example that drives O-C code and computes standard
%  statistical products typical of O-C analysis.  This script is an example
%  of how R-M might use the O-C data.

clear all; close all; clc;
showPlots = false;
% Create the epoch object and set the initial epoch
date = AbsoluteDate();
date.SetGregorianDate(2017,1,15,22,30,20.111);

% Create the spacecraft state object and set Keplerian elements
state = OrbitState();
state.SetKeplerianState(6700,0.002,90*pi/180,pi/4 + pi/6,.2345,pi/6);

% Create a conical sensor
sensor = ConicalSensor(pi);

% Create a spacecraft giving it a state and epoch
sat1 = Spacecraft(date,state);
sat1.AddSensor(sensor)

% Create the propagator
prop = Propagator(sat1);

% Create the point group and initialize the coverage checker
pGroup = PointGroup();
pGroup.AddHelicalPointsByNumPoints(200);
covChecker = CoverageChecker(pGroup,sat1);
if showPlots
    figHandle = pGroup.PlotAllTestPoints();
end

% Propagate for a duration and collect data
startDate = date.GetJulianDate();
count = 0;
while date.GetJulianDate() < startDate + 1
    
    % Propagate
    date.Advance(120);
    prop.Propagate(date);
    
    % Compute points in view
    accessPoints = covChecker.AccumulateCoverageData();
    
    %% Plotting stuff used for testing and visualization
    if showPlots
        % Save orbit for plotting. Kludge cause output classes not done yet.
        count = count + 1;
        intertialState = sat1.GetCartesianState();
        jDate = sat1.GetJulianDate();
        cb = Earth();
        orbitState(count,:) = cb.GetBodyFixedState(intertialState(1:3,1),jDate)';
        pGroup.PlotSelectedTestPoints(accessPoints); drawnow;
        if count >= 2
            figure(figHandle); hold on;
            plot3(orbitState(count-1:count,1)/6378,orbitState(count-1:count,2)/6378,...
                orbitState(count-1:count,3)/6378,'k-','MarkerSize',3)
        end
    end
    
end

% Compute coverage data
coverageEvents = covChecker.ProcessCoverageDate();

%%  Compute coverate stats.  Shows how R-M might use data for coverage analysis

% Create Lat\Lon Grid
for pointIdx = 1:pGroup.GetNumPoints()
    vec = pGroup.GetPointPositionVector(pointIdx);
    lonVec(pointIdx,1) = atan2(vec(2),vec(1))*180/pi;
    latVec(pointIdx,1) = asin(vec(3)/norm(vec))*180/pi;
end

% Compute total coverage statistics from all coverage events
totalCoverageDuration = zeros(pGroup.GetNumPoints(),1);
numPassVec = zeros(pGroup.GetNumPoints(),1);
minPassVec = numPassVec;
maxPassVec = numPassVec;
for eventIdx = 1:length(coverageEvents)
    currEvent = coverageEvents{eventIdx};
    poiIndex = currEvent.poiIndex;
    eventDuration = (currEvent.endDate.GetJulianDate() - currEvent.startDate.GetJulianDate())*24;
    totalCoverageDuration(poiIndex) = totalCoverageDuration(poiIndex) + eventDuration;
    vec = pGroup.GetPointPositionVector(pointIdx);
    %     lat = atan2(vec(2),vec(1))*180/pi;
    %     lon = asin(vec(3)/norm(vec));
    %     Z(lat,lon) = totalCoverageDuration;
    
    % Save the maximum duration if necessary
    if eventDuration > maxPassVec(poiIndex)
        maxPassVec(poiIndex) = eventDuration;
    end
    
    if minPassVec(poiIndex) == 0 ||  (eventDuration< maxPassVec(poiIndex))
        minPassVec(poiIndex) = eventDuration;
    end
    numPassVec(poiIndex) = numPassVec(poiIndex) + 1;
end



%%  Write the simple coverage report to the MATLAB command window
disp('       =======================================================================')
disp('       ==================== Brief Coverage Analysis Report ===================')
disp('       lat (deg): Latitude of point in degrees                  ')
disp('       lon (deg): Longitude of point in degrees                  ')
disp('       numPasses: Number of total passes seen by a point                           ')
disp('       totalDur : Total duration point was observed in hours                         ');
disp('       minDur   : Duration of the shortest pass in minutes                         ');
disp('       maxDur   : Duration of the longest pass in hours                            ')
disp('       =======================================================================')
disp('       =======================================================================')
disp('  ')

format short g
data = [latVec,lonVec, numPassVec, totalCoverageDuration, minPassVec, maxPassVec];
headerCount = 1;
for passIdx = 1:10:pGroup.GetNumPoints()
    disp('       lat (deg)    lon (deg)     numPasses   totalDur    minDur      maxDur')
    dataEnd = passIdx + 9;
    disp(data(passIdx:dataEnd,:))
end
if dataEnd + 1 < pGroup.GetNumPoints()
 disp('       lat (deg)    lon (deg)     numPasses   totalDur    minDur      maxDur')
 disp(data(dataEnd + 1 :end,:))
end

figure(2)
subplot(3,1,1)
xlin= -180:5:180;
ylin= -90:5:90;
[X,Y]=meshgrid(xlin,ylin);
Z=griddata(lonVec,latVec,maxPassVec*60,X,Y,'cubic');
h = contour(X,Y,Z);
clabel(h);
grid on
subplot(3,1,2)
xlin= -180:5:180;
ylin= -90:5:90;
[X,Y]=meshgrid(xlin,ylin);
Z=griddata(lonVec,latVec,minPassVec*60,X,Y,'cubic');
h = contour(X,Y,Z);
clabel(h);
grid on
subplot(3,1,3)
xlin= -180:5:180;
ylin= -90:5:90;
[X,Y]=meshgrid(xlin,ylin);
Z=griddata(lonVec,latVec,totalCoverageDuration*60,X,Y,'cubic');
h = contour(X,Y,Z);
clabel(h);
grid on

File: prototype/TAT-C/cpp/test/SystemTest_Analysis.cpp
//------------------------------------------------------------------------------
//                               SystemTest_Analysis
//------------------------------------------------------------------------------
// GMAT: General Mission Analysis Tool
//
// Author: Wendy Shoan
// Created: 2016.05.31
//
/**
 * System tester for analysis
 */
//------------------------------------------------------------------------------

#include <iostream>
#include <string>
#include <ctime>
#include <cmath>
#include "gmatdefs.hpp"
#include "GmatConstants.hpp"
#include "Rvector6.hpp"
#include "Rvector3.hpp"
#include "Rmatrix.hpp"
#include "RealUtilities.hpp"
#include "MessageInterface.hpp"
#include "ConsoleMessageReceiver.hpp"
#include "AbsoluteDate.hpp"
#include "Spacecraft.hpp"
#include "Earth.hpp"
#include "KeyValueStatistics.hpp"
#include "VisiblePOIReport.hpp"
#include "OrbitState.hpp"
#include "PointGroup.hpp"
#include "Propagator.hpp"
#include "ConicalSensor.hpp"
#include "CoverageChecker.hpp"
#include "TimeTypes.hpp"


using namespace std;
using namespace GmatMathUtil;
using namespace GmatMathConstants;

//------------------------------------------------------------------------------
// int main(int argc, char *argv[])
//------------------------------------------------------------------------------
int main(int argc, char *argv[])
{
   std::string outFormat = "%16.9f ";
   Real        tolerance = 1e-15;
   
   ConsoleMessageReceiver *consoleMsg = ConsoleMessageReceiver::Instance();
   MessageInterface::SetMessageReceiver(consoleMsg);
   std::string outPath = "./";
   MessageInterface::SetLogFile(outPath + "GmatLog.txt");
   MessageInterface::ShowMessage("%s\n",
                                 GmatTimeUtil::FormatCurrentTime().c_str());
   
   // Set global format setting
   GmatGlobal *global = GmatGlobal::Instance();
   global->SetActualFormat(false, false, 16, 1, false);
   
   char *buffer = NULL;
   buffer = getenv("OS");
   if (buffer  != NULL)
   {
      MessageInterface::ShowMessage("Current OS is %s\n", buffer);
   }
   else
   {
      MessageInterface::ShowMessage("Buffer is NULL\n");
   }
   
   MessageInterface::ShowMessage("*** START TEST ***\n");
   
   
   try
   {
      // Test the PointGroup
      MessageInterface::ShowMessage("*** TEST*** Analysis!!!!\n");
      
      // This is a usage example that drives O-C code and computes standard
      //  statistical products typical of O-C analysis.  This script is an example
      //  of how R-M might use the O-C data.
      
      AbsoluteDate  *date;
      OrbitState    *state;
      ConicalSensor *sensor;
      Spacecraft    *sat1;
      Propagator    *prop;
      PointGroup    *pGroup;
      std::vector<VisiblePOIReport> coverageEvents;

      
      clock_t t0 = clock();
      Integer numIter = 1;

      for (Integer ii = 0; ii < numIter; ii++) // **********
      {
      
      bool showPlots = false;

      // Create the epoch object and set the initial epoch
      date = new AbsoluteDate();
      date->SetGregorianDate(2017,1,15,22,30,20.111);
      
//      MessageInterface::ShowMessage(" --- date created\n");
      
      // Create the spacecraft state object and set Keplerian elements
      state = new OrbitState();
      state->SetKeplerianState(6700.0,0.002,90.0*RAD_PER_DEG,
                               PI/4.0 + PI/6.0,0.2345,PI/6.0);
      
//      MessageInterface::ShowMessage(" --- state created\n");
      
      // Create a conical sensor
      sensor = new ConicalSensor(PI);
      
//      MessageInterface::ShowMessage(" --- conical sensor created\n");
      
      // Create a spacecraft giving it a state and epoch
      sat1 = new Spacecraft(date,state);
      sat1->AddSensor(sensor);
      
//      MessageInterface::ShowMessage(" --- spacecraft created\n");
      
      // Create the propagator
      prop = new Propagator(sat1);
      
//      MessageInterface::ShowMessage(" --- propagator created\n");
      
      // Create the point group and initialize the coverage checker
      pGroup = new PointGroup();
      pGroup->AddHelicalPointsByNumPoints(200);
      CoverageChecker *covChecker = new CoverageChecker(pGroup,sat1);
      
//      MessageInterface::ShowMessage(" --- point group created\n");

      if (showPlots)
         ; // don't have this in C++
//         figHandle = pGroup.PlotAllTestPoints();
  
      // Propagate for a duration and collect data
      Real    startDate = date->GetJulianDate();
      Integer count     = 0;
      // Over 1 day
      while (date->GetJulianDate() < ((Real)startDate + 1.0))
      {
//         MessageInterface::ShowMessage(" --- about to advance the time and propagate\n");
         // Propagate
         date->Advance(120.0);
         prop->Propagate(*date);
//         MessageInterface::ShowMessage(" --- done advancing the time and propagating\n");
      
         // Compute points in view
         IntegerArray accessPoints = covChecker->AccumulateCoverageData();
      
         // Plotting stuff used for testing and visualization
         if (showPlots)
         {
            ;
//            // Save orbit for plotting. Kludge cause output classes not done yet.
//            count++;
//            Rvector6 intertialState = sat1->GetCartesianState();
//            Real     jDate          = sat1.GetJulianDate();
//            Earth *cb = new Earth();
//            orbitState(count,:) = cb.GetBodyFixedState(intertialState(1:3,1),jDate)';
//            pGroup.PlotSelectedTestPoints(accessPoints); drawnow;
//            if count >= 2
//               figure(figHandle); hold on;
//            plot3(orbitState(count-1:count,1)/6378,orbitState(count-1:count,2)/6378,...
//                  orbitState(count-1:count,3)/6378,'k-','MarkerSize',3)
         }
      }
      
//      MessageInterface::ShowMessage(" --- propagation completed\n");
      
      // Compute coverage data
      coverageEvents = covChecker->ProcessCoverageData();
      
//      MessageInterface::ShowMessage(" --- ProcessCoverageData completed (%d reports)\n",
//                                    (Integer) coverageEvents.size());

      } // *****
      Real timeSpent = ((Real) (clock() - t0)) / CLOCKS_PER_SEC;
      MessageInterface::ShowMessage("TIME SPENT in %d iterations is %12.10f seconds\n",
                                    numIter,timeSpent);

      
      RealArray lonVec;
      RealArray latVec;
      // Compute coverate stats.  Shows how R-M might use data for coverage analysis
      // Create Lat\Lon Grid
      for (Integer pointIdx = 0; pointIdx < pGroup->GetNumPoints(); pointIdx++)
      {
         Rvector3 *vec = pGroup->GetPointPositionVector(pointIdx);
         lonVec.push_back(ATan(vec->GetElement(1),vec->GetElement(0))*DEG_PER_RAD);
         latVec.push_back(ASin(vec->GetElement(2)/vec->GetMagnitude())*DEG_PER_RAD);
      }
      
      MessageInterface::ShowMessage(" --- lat/long set-up completed\n");
      
      // Compute total coverage statistics from all coverage events
      Rvector totalCoverageDuration(pGroup->GetNumPoints());
      IntegerArray numPassVec;  // (pGroup->GetNumPoints());
      for (Integer ii = 0; ii < pGroup->GetNumPoints(); ii++)
         numPassVec.push_back(0);
      Rvector minPassVec(pGroup->GetNumPoints());
      Rvector maxPassVec(pGroup->GetNumPoints());
      for (Integer eventIdx = 0; eventIdx < coverageEvents.size(); eventIdx++)
      {
         VisiblePOIReport currEvent      = coverageEvents.at(eventIdx);
         Integer          poiIndex       = currEvent.GetPOIIndex();
         Real             eventDuration  = (currEvent.GetEndDate().GetJulianDate() -
                                            currEvent.GetStartDate().GetJulianDate()) * 24.0;
         totalCoverageDuration(poiIndex) = totalCoverageDuration(poiIndex) + eventDuration;
//         Rvector3 vec = pGroup->GetPointPositionVector(pointIdx); // ?
         
//      %     lat = atan2(vec(2),vec(1))*180/pi;
//      %     lon = asin(vec(3)/norm(vec));
//      %     Z(lat,lon) = totalCoverageDuration;
      
      // Save the maximum duration if necessary
         if (eventDuration > maxPassVec(poiIndex))
            maxPassVec(poiIndex) = eventDuration;

      
         if (minPassVec(poiIndex) == 0 ||  (eventDuration< maxPassVec(poiIndex)))
            minPassVec(poiIndex) = eventDuration;

         numPassVec.at(poiIndex) = numPassVec.at(poiIndex) + 1;
      }
      
      // *********** display stuff ***********
      // Write the simple coverage report to the MATLAB command window
      MessageInterface::ShowMessage("       =======================================================================\n");
      MessageInterface::ShowMessage("       ==================== Brief Coverage Analysis Report ===================\n");
      MessageInterface::ShowMessage("       lat (deg): Latitude of point in degrees                  \n");
      MessageInterface::ShowMessage("       lon (deg): Longitude of point in degrees                  \n");
      MessageInterface::ShowMessage("       numPasses: Number of total passes seen by a point                           \n");
      MessageInterface::ShowMessage("       totalDur : Total duration point was observed in hours                         \n");
      MessageInterface::ShowMessage("       minDur   : Duration of the shortest pass in minutes                         \n");
      MessageInterface::ShowMessage("       maxDur   : Duration of the longest pass in hours                            \n");
      MessageInterface::ShowMessage("       =======================================================================\n");
      MessageInterface::ShowMessage("       =======================================================================\n");
      MessageInterface::ShowMessage("  ");
      
//      data = [latVec,lonVec, numPassVec, totalCoverageDuration, minPassVec, maxPassVec];  <<<<<<<<<
      Integer headerCount = 1;
      Integer dataEnd     = 0;
      for (Integer passIdx = 0; passIdx < pGroup->GetNumPoints(); passIdx+= 10)
      {
         MessageInterface::ShowMessage("       lat (deg)     lon (deg)       numPasses  totalDur    minDur      maxDur\n");
         dataEnd = passIdx + 10;  // 9;
         for (Integer ii = 0; ii < 10; ii++)  // 9
            MessageInterface::ShowMessage("       %le    %le    %d    %le    %le    %le \n",
//            MessageInterface::ShowMessage("       %12.10f    %12.10f    %d    %12.10f    %12.10f    %12.10f \n",
                                          latVec.at(passIdx+ii),
                                          lonVec.at(passIdx+ii),
                                          numPassVec.at(passIdx+ii),
                                          totalCoverageDuration(passIdx+ii),
                                          minPassVec(passIdx+ii),
                                          maxPassVec(passIdx+ii));

      }

      if (dataEnd + 1 < pGroup->GetNumPoints())
      {
         MessageInterface::ShowMessage("       lat (deg)    lon (deg)     numPasses   totalDur    minDur      maxDur\n");
         for (Integer ii = dataEnd; ii < pGroup->GetNumPoints(); ii++)
            MessageInterface::ShowMessage("       %le    %le    %d    %le    %le    %le \n",
//            MessageInterface::ShowMessage("       %12.10f    %12.10f    %d    %12.10f    %12.10f    %12.10f \n",
                                          latVec.at(ii),
                                          lonVec.at(ii),
                                          numPassVec.at(ii),
                                          totalCoverageDuration(ii),
                                          minPassVec(ii),
                                          maxPassVec(ii));
      }

      
      cout << endl;
      cout << "Hit enter to end" << endl;
      cin.get();
      
      // delete pointers here
      
      MessageInterface::ShowMessage("*** END TEST ***\n");
   }
   catch (BaseException &be)
   {
      MessageInterface::ShowMessage("Exception caught: %s\n", be.GetFullMessage().c_str());
   }
   
}
